cmd/outline-ss-server/main.go

```diff
@@ -122,65 +111,81 @@ func (s *SSServer) loadConfig(filename string) error {
 		return fmt.Errorf("failed to load config (%v): %w", filename, err)
 	}
 
-	portChanges := make(map[int]int)
-	portCiphers := make(map[int]*list.List) // Values are *List of *CipherEntry.
-	for _, keyConfig := range config.Keys {
-		portChanges[keyConfig.Port] = 1
-		cipherList, ok := portCiphers[keyConfig.Port]
+	// We hot swap the config by having the old and new listeners both live at
+	// the same time. This means we create listeners for the new config first,
+	// and then close the old ones after.
+	oldListenerSet := s.lnSet
+	s.lnSet = s.lnManager.NewListenerSet()
+	var totalCipherCount int
+
+	portHandlers := make(map[int]Handler)
```

Comment:
I think the right abstraction here is port -> Service, not Handler.

- for each key:
  - create the Service for the port, if it doesn't exist
  - add the key to the Service

Then you can go over the list of services and start each listener its serving loop.